build/build.sh

```diff
@@ -207,6 +219,19 @@ edge_dir=${EDGE_DIR:-Edge}
 # Disable with '-disable-werror' option.
 cmake_flags+=" -DXRT_ENABLE_WERROR=$werror"
 
+if [[ $split_static == 1 ]]; then
+    cmake_flags+=" -DXRT_STATIC_COMPONENT=xrt-static"
+
+    if [[ $split_dev == 1 ]]; then
+	cmake_flags+=" -DXRT_DEV_COMPONENT=xrt-dev"
+    fi
```

Comment:
The dynamic libraries are useful without the headers if you are only interested in deploying an already compiled host application.  Most of the system package managers in Linux distributions ship a runtime package that only contains executables and dynamic-libs and a split 'development' package that contains headers and other things needed to link against the libraries.

I agree that the static libraries are useless without the headers and that's why the `xrt-static` package would depend on the `XRT_DEV_COMPONENT` package that contains the headers.  The intent is to allow users who do not want to ever accidentally link the static version of the library to install only the dynamic libs.

Reply:
> The dynamic libraries are useful without the headers if you are only interested in deploying an already compiled host application. Most of the system package managers in Linux distributions ship a runtime package that only contains executables and dynamic-libs and a split 'development' package that contains headers and other things needed to link against the libraries.

Yes, this matches what I mentioned.

> I agree that the static libraries are useless without the headers and that's why the `xrt-static` package would depend on the `XRT_DEV_COMPONENT` package that contains the headers. The intent is to allow users who do not want to ever accidentally link the static version of the library to install only the dynamic libs.

I am not convinced that we need a static package.  AFAIK the *-dev packages always contains link libraries whether static or dynamic and headers.  I don't know about conda, but I know that XRT does not package for different environments, and I would hate to see Conda specific changes to XRT.

Reply:
> I am not convinced that we need a static package. AFAIK the *-dev packages always contains link libraries whether static or dynamic and headers. I don't know about conda, but I know that XRT does not package for different environments, and I would hate to see Conda specific changes to XRT. The accidental linking against dynamic vs static, is a pilot error.

@stsoe thanks for looking at this!

At the end of the day, if you don’t want to support separation of dynamic and static libraries, It’s your call, and we will respect it.

I just want to state that there are established conventions that most projects use, where -DBUILD_SHARED_LIBS=ON controls whether static or dynamic libraries are built. Protobuf, glog, gflags, json-c all obey. Their users may not want to waste disk space on static libraries if they will never use them. It really is not a conda specific request. Perhaps we should see how they implement this feature, as it may be simpler than adding a new component.

Reply:
> At the end of the day, if you don’t want to support separation of dynamic and static libraries, It’s your call, and we will respect it.

completely agree.

Reply:
Yeah, I just don't believe in the disk space issue for developers.  I am absolutely in agreement regarding making runtime package as small as possible, but I have never seen the static library vs dynamic link libraries split you are talking about.  It really would require a total of 4 package for this to work

1. runtime
2. headers (dev)
3. dynamic link (dev)
4. static link (dev)

For a project the size of XRT this is an overkill IMO and in fact I don't recall ever seeing this kind of split in any packages I install with apt.   

Can we please just enable a runtime package and a development package?  Enabling this is long over due and I really appreciate you bringing this back to the surface for discussion and action.

Reply:
I think that the dynamic link libraries are needed in runtime, along with any executables.  

You're right about debian and ubuntu not having this split.  I was curious what I could find about this topic outside of the conda ecosystem, so I did a little digging, just for my own entertainment:
* Debian
  * Debian policy says put the static libs in the dev package: https://www.debian.org/doc/debian-policy/ch-sharedlibs.html#static-libraries
  * However, they also have a page that talks about the downsides of static linking and how to mitigate them: https://wiki.debian.org/StaticLinking
    * they have a tool that identifies and tracks statically linked binaries in their packages https://lintian.debian.org/tags/statically-linked-binary
    * they also keep source code around for licensing reasons.  Since everything in XRT that ends up in a static library is Apache 2, perhaps this isn't an issue.  I'm still unclear as to exactly what in XRT gets infected with GPL, probably only the kernel modules but I personally haven't reviewed it.  Also, perhaps the availability of the source code on github satisfies the GPL requirements.  Again, I'm neither a licensing expert nor a lawyer.
 
* OpenSUSE
  * doesn't explicitly allow or disallow them but [says it is best practice to avoid them](https://en.opensuse.org/openSUSE:Shared_library_packaging_policy#exception_4:~:text=Avoid%20packaging%20static%20libraries.%20You%20should%20use%20%2D%2Ddisable%2Dstatic%20configure%20option%20or%2C%20as%20a%20last%20resort%2C%20remove%20static%20libraries%20after%20make%20install.%20If%20in%20doubt%2C%20ask.)
* Fedora
  * https://docs.fedoraproject.org/en-US/packaging-guidelines/#packaging-static-libraries starts "Packages including libraries SHOULD exclude static libs as far as possible"
  * Requires shipped static libs to be identified with a `-static` package (but allow it to be a 'virtual' package) and any users of the static libraries must declare their dependence on the `-static` package.  I thought this was an interesting idea and scanned through https://github.com/conda-forge/cfep/pull/34 (adopting cfep-18) to see if it was discussed.  I don't see it there.  I suspect that allowing the virtual `-static` package would be difficult in practice to enforce without support in the tooling.  I think cfep-18 takes a KISS approach by requiring the static-libs to be in a separate package
  
Got tired of looking right now but it might be interesting to see if Cannonical or RedHat have specific statements regarding the packaging of static libs.
  
  
I will acknowledge that XRT probably has one of the best use-cases for wanting to support statically linked executables by it's end users.  I would think most people that bundle their host executable code into the xclbin will want it to be statically linked so that they don't have to worry about keeping track of all the dynamic libs for deploying an app, it's just all bundled up nicely in the xclbin.

Reply:
> I think that the dynamic link libraries are needed in runtime, along with any executables.

No, definitely not.   On Linux, the link libraries are softlinks installed by the development package.  The softlinks are named `foo.so` and point to a versioned library (foo.so -> foo.so.2 -> foo.so.2.1.0) installed by the runtime package.

Reply:
Oh, ok, I understand what you were saying now.  The unversioned symbolic links to the dynamic libraries definitely do need to only be installed in the dev package.  I was confused and thought you meant all dynamic library artifacts.